app/lib/RouteResolver.js

```diff
@@ -120,23 +153,46 @@ RouteResolver.prototype.register = function register(route) {
     route.response = handler.call(this, route.response);
   }
 
+  route.pathname = parse(route.path, true).pathname;
+  route.pathRegExp = pathToRegExp(route.pathname);
+
   const expectation = new Expectation(route);
+  route.expectation = expectation;
 
   // unregister existing matching routes
   this.unregister([route]);
 
-  this.app[route.method](route.path, expectation.middleware.bind(expectation), route.response);
+  this.routes.push(route);
+
+  this.listen();
 
   return {
     expect: () => expectation.api()
   };
 };
 
 RouteResolver.prototype.unregister = function unregister(routes) {
-  this.app._router.stack = this.app._router.stack.filter((layer) => {
-    return !(layer.route && routes.some((route) => {
-      return route.path === layer.route.path && layer.route.methods[route.method] === true;
-    }));
+  this.routes = this.routes.filter(oldRoute =>
+    !routes.some(newRoute => isRouteToReplace(newRoute, oldRoute))
```

Comment:
Filter out any old routes that match any new of the new routes being unregistered.